test/core/ticktree/TickTreeMarketOrder.t.sol

```diff
@@ -0,0 +1,243 @@
+                for (uint hs = 1; hs < tickListScenarios.length; ++hs) {
+                  scenario.higherTick.size = tickListScenarios[hs][0];
+                  scenario.higherTick.offersToTake = tickListScenarios[hs][1];
+                  run_market_order_scenario(scenario, printToConsole);
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+
+    vm.resumeGasMetering();
+  }
+
+  // function test_single_scenario() public {
+  //   run_market_order_scenario(
+  //     MarketOrderScenario({tickScenario: TickScenario({tick: Tick.wrap(0), hasHigherTick: true, higherTick: 4, hasLowerTick: false, lowertick: Tick.wrap(0)}), offerTickListSize: 1, offerPos: 0}),
+  //     true
+  //   );
+  // }
+
+  function scenarioToString(TickListScenario memory scenario) internal pure returns (string memory) {
+    string memory tickListScenario = scenario.size == 0
+      ? "empty          "
+      : scenario.size == 1 ? "fully taken    " : scenario.offersToTake == 1 ? "partially taken" : "not taken      ";
+    return string.concat(
+      tickListScenario,
+      ", tick: ",
+      toString(scenario.tick),
+      ", size: ",
+      vm.toString(scenario.size),
+      ", offersToTake: ",
+      vm.toString(scenario.offersToTake)
+    );
+  }
+
+  function removeTakenOffers(TickTree storage tickTree, TickListScenario memory scenario, uint[] memory offerIds)
+    internal
+  {
+    for (uint i = 0; i < scenario.offersToTake; ++i) {
+      removeOffer(tickTree, offerIds[i]);
+    }
+  }
+
+  function run_market_order_scenario(MarketOrderScenario memory scenario, bool printToConsole) internal {
+    if (printToConsole) {
+      console.log("market order scenario");
+      console.log("  lower tick scenario:  ", scenarioToString(scenario.lowerTick));
+      console.log("  middle tick scenario: ", scenarioToString(scenario.middleTick));
+      console.log("  higher tick scenario: ", scenarioToString(scenario.higherTick));
+    }
+
+    // 1. Capture state before test
+    uint vmSnapshotId = vm.snapshot();
+
+    // 2. Create scenario
+    (uint[] memory lowerOfferIds, uint lowerOffersGive) =
+      add_n_offers_to_tick(Tick.unwrap(scenario.lowerTick.tick), scenario.lowerTick.size);
+    (uint[] memory middleOfferIds, uint middleOffersGive) =
+      add_n_offers_to_tick(Tick.unwrap(scenario.middleTick.tick), scenario.middleTick.size);
+    (uint[] memory higherOfferIds, uint higherOffersGive) =
+      add_n_offers_to_tick(Tick.unwrap(scenario.higherTick.tick), scenario.higherTick.size);
+    uint fillVolume = lowerOffersGive * scenario.lowerTick.offersToTake
+      + middleOffersGive * scenario.middleTick.offersToTake + higherOffersGive * scenario.higherTick.offersToTake;
+
+    // 3. Snapshot tick tree
+    TickTree storage tickTree = snapshotTickTree();
+
+    // 4. Run the market order
+    mgv.marketOrderByLogPrice(olKey, MAX_LOG_PRICE, fillVolume, true);
```

Comment:
Consider (but maybe separate PR) whether it is worth it to snapshot the resulting ticktree inside posthook of the last taken offer, so that we verify tree is right already when posthooks are executed. Not high priority.

Reply:
Good idea.  I've created an issue on it: https://github.com/mangrovedao/mangrove-core/issues/506